### ISSUE TITLE
Should clause heads do matching before unification?

### ISSUE THREAD

**Issue by mb64** [LPCIC/elpi#123]
```prolog
count-down-from 0 Xs Xs.
count-down-from N [N|Xs] Xs1 :- N1 is N - 1, count-down-from N1 Xs Xs1.
```
This predicate has a `(+,-,+)` mode, but adding this type/mode declaration for `count-down-from` makes it no longer work:
```prolog
pred count-down-from i:int, o:(list int), i:(list int).
```
This is because in the first clause, it first unifies its second argument with the fresh unification var `Xs`, then matches `Xs` against the third argument, and never actually unifies the two.

Would it make sense to have ELPI reorder how it matches clause heads to do matching before it does unification? It would make this particular edge case more well-behaved, but might cause confusing behavior in other edge cases.

**Comment by gares** [comment]
That is a technical eye! Indeed using `i:` mode and non-linear patterns is not very well defined.
But I'm not so sure I get your code, this is how I'm running it:
```prolog
pred count-down-from i:int, o:(list int), i:(list int).
count-down-from 0 Xs Xs.
count-down-from N [N|Xs] Xs1 :- N1 is N - 1, count-down-from N1 Xs Xs1.

main :-
  count-down-from 3 X [2,3], print X.
```
and I get `[3,2,1,2,3]` which seems correct.

**Comment by mb64** [comment]
You're right, it does work! I'm sorry, I should've checked my example before asking.

In my actual code, I had
```prolog
pred fill-metas i:(int -> k -> ty), i:ty, i:int, o:int, o:(list k), i:(list k).
% ... some clauses
fill-metas _ _ N N Ks Ks.
fill-metas _ _ _ _ _ _ :- halt "last clause didn't match".
```
My program would error with the message `last clause didn't match`, but changing the last real clause to
```prolog
fill-metas _ _ N N Ks Ks1 :- Ks = Ks1.
```
fixed it.

I will try to reduce it to a minimal example that actually demonstrates the behavior.

**Comment by gares** [comment]
I think the "bug" is there, but I never had a real use case for it. Looking forward to your real code then. You could also try changing the order of the arguments, and see if this is acceptable. I'm not in love with the idea of making the operational semantics harder to predict. It would be nicer to detect the problematic rule statically if possible and suggest some work around.

**Comment by mb64** [comment]
Trying to minimize my example, I've figured out that the last clause *was* in fact matching, but a mistake in a later part of the program made it backtrack to the erroring clause.  And at the same time as I changed it use `Ks = Ks1`, I made a minor change elsewhere which I thought was unrelated, but turned out to be the real thing that fixed my bug!  How embarrassing.

Sorry for the mistaken bug report! Feel free to close it.
